pdata/internal/cmd/pdatagen/internal/base_slices.go

```diff
@@ -17,17 +17,25 @@ const sliceTemplate = `// {{ .structName }} logically represents a slice of {{ .
 // Important: zero-initialized instance is not valid for use.
 type {{ .structName }} struct {
 	orig *[]{{ .originElementType }}
+	state *internal.State
 }
 
-func new{{ .structName }}(orig *[]{{ .originElementType }}) {{ .structName }} {
-	return {{ .structName }}{orig}
+func new{{ .structName }}(orig *[]{{ .originElementType }}, state *internal.State) {{ .structName }} {
+	return {{ .structName }}{orig: orig, state: state}
 }
 
 // New{{ .structName }} creates a {{ .structName }} with 0 elements.
 // Can use "EnsureCapacity" to initialize with a given capacity.
 func New{{ .structName }}() {{ .structName }} {
 	orig := []{{ .originElementType }}(nil)
-	return new{{ .structName }}(&orig)
+	state := internal.StateMutable
+	return new{{ .structName }}(&orig, &state)
+}
+
+func (es {{ .structName }}) assertMutable() {
+	if *es.state == internal.StateReadOnly {
+		panic("invalid access to shared data")
+	}
```

Comment:
Instead of making every struct to have this assertMutable it is easier to make assertMutable on the state object so you will call `es.state.assertMutable()`

Reply:
Good idea. I updated the code